keps/sig-storage/20200427-csi-driver-skip-permission.md

```diff
@@ -0,0 +1,155 @@
+- [Alternatives [optional]](#alternatives-optional)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+Currently before bind mounting a CSI volume for a Pod, we use imprecise heuristics,
+such as presence of fsType on the PVC to determine if the volume supports fsGroup based
+permission change. These heuristics are known to be fragile, and cause problems with different
+storage types.
+
+To solve this issue we will add a new field called `CSIDriver.Spec.SupportsFSGroup` 
+that allows the driver to define if it supports volume ownership modifications via
+fsGroup.
+
+## Motivation
+
+When a volume is mounted on the node, we modify volume ownership 
+before bind mounting the volume inside container; however, if the volume
+does not support these operations (such as NFS), then the change is still attempted. 
+This results in errors being reported to the user if the volume doesn't 
+support these operations.
+
+### Goals
+
+ - Allow CSIDrivers to opt-in to volume ownership changes.
+
+### Non-Goals
+
+ - This does not attempt to update existing CSIDrivers themselves.
+
+## Proposal
+
+We propose that the `CSIDriver` type include a field that defines if the volume 
+provided by the driver supports changing volume ownership. This will be enabled
+with a new feature gate, `CSIVolumeSupportFSGroup`.   
+
+### Implementation Details/Notes/Constraints [optional]
+
+Currently Volume permission and ownership change is done automatically for every volume. 
+As part of this proposal we will change the algorithm that modifies volume ownership and permissions
+for CSIDrivers to check the new field, and skip volume ownership modifications if it is found to be
+`false`.
+
+When defining a `CSIDriver`, we propose that `CSIDriver.Spec` be expanded to include a new field entitled 
+`SupportsFSGroup` which can have following possible values:
+
+ - `nil` --> Current behavior. Attempt to modify the volume ownership and permissions to the defined `fsGroup` when the volume is attached.
+ - `false` --> New behavior. Attach the volume without attempting to modify volume ownership or permissions.
+ - `true` --> New behavior. Always attempt to apply the defined fsGroup to modify volume ownership and permissions.
+
+```go
+type CSIDriverFSGroupChangePolicy string
+
+type CSIDriverSpec struct {
+    // SupportsFSGroup ← new field
+    // Defines if the underlying volume supports changing ownership and 
+    // permission of the volume before being attached. 
+    // If set to true, SupportsFSGroup indicates that 
+    // the underlying volume supports volume ownership and permission
+    // changes, and the filesystem will be modified by the defined fsGroup.
+    // If set to false, then the volume will be attached without modifying
+    // the volume's ownership or permissions.
+    // Defaults to nil, which results in the volume being examined
+    // and the volume ownership and permissions attempting to be updated
```

Comment:
Can we be more explicit about what this behavior is? It's not just fsGroup, it's accessMode too that determines this.

Reply:
I've updated this to include information about accessModes as well.

Reply:
I think we can be even more explicit and say that it's only modified if fsGroup is set and accessMode is RWO.